### PR TITLE
Draft: Set subnetwork properties on instance instead of the class

### PR DESCRIPTION
**Pull request by eendebakpt**
After importing skrf and creating a 1-port network, the network as attributes `s13', `s3_3` which it should not have (since it is a 1-port network). A minimal example:
```
import numpy as np
import skrf as rf

f = rf.Frequency.from_f(np.linspace(2e6, 3e6, 10), unit="Hz")
self=n=rf.Network(frequency=f, s=np.random.rand(10,)*1j, name='test')
# now n has properties s_12, which it should not have

w=[p for p in dir(self) if p.startswith('s') and p[-1] in ['0', '1','2','3','4']]
print(w)
```
has output
```
['s11', 's12', 's13', 's1_1', 's1_2', 's1_3', 's21', 's22', 's23', 's2_1', 's2_2', 's2_3', 's31', 's32', 's33', 's3_1', 's3_2', 's3_3', 's_db10']
```

This PR fixes that by setting the attributes on the instance, instead of the class.

### CODE REVIEW

**Comment by FranzForstmayr** [comment]
Nice catch!

**Comment by eendebakpt** [comment]
@FranzForstmayr This is more complicated than I expected. Setting a property on an instance is not really possible. I used some code from stackoverflow to make this work (see commit 
fc45854), but now the equality checks are failing. I could try to fix that, but we risk adding all kinds of non-standard workarounds.

An alternative option is to subclass `Network` into `Network1port`, `Network2port`, ..., `NetworkManyPort` and set the appropriate properties for each subclass.

And a mitigation could be to modify the `__dir__` of Network to filter out values of `sm_n` that are not valid for the number of ports.

File: skrf/frequency.py [hunk 1]
```diff
@@ -208,7 +208,6 @@ def __getitem__(self, key: Union[str, int, slice]) -> 'Frequency':
         if isinstance(key, str):
 
             # they passed a string try and do some interpretation
-            re_numbers = re.compile(r'.*\d')
```
Comment (jhillairet) [inline skrf/frequency.py:211]: Why does this line have been removed? Wasn't used after?

**Comment by jhillairet** [comment]
> An alternative option is to subclass `Network` into `Network1port`, `Network2port`, ..., `NetworkManyPort` and set the appropriate properties for each subclass.

I'm not a big fan of this solution.
 
>  filter out values of `sm_n` that are not valid for the number of ports.

Removing properties `smn` and `sm_n` for (m,n) higher than the number of ports in `__generate_subnetworks` should solve the issue.

File: skrf/frequency.py (hunk 1 above)
Comment (eendebakpt) [reply skrf/frequency.py:211]: No, it was not used in the file

**Comment by eendebakpt** [comment]
> > An alternative option is to subclass `Network` into `Network1port`, `Network2port`, ..., `NetworkManyPort` and set the appropriate properties for each subclass.
> 
> I'm not a big fan of this solution.
Me neither.
> 
> > filter out values of `sm_n` that are not valid for the number of ports.
> 
> Removing properties `smn` and `sm_n` for (m,n) higher than the number of ports in `__generate_subnetworks` should solve the issue.
Removing the properties for high (m,n) in `__generate_subnetworks` does not work: the properties are set on the `Network` class, and not on the different `Network` instances. Since different `Network` instances can have different values of `m` and `n`, we will get incorrect `s` properties as soon as we have `Network` instances with different number of ports.

Since `dir` works on the instance we can modify that. The incorrect `smn` parameters are still part of the object, but they won't show any more in auto-completion. It is only a (partial) mitigation, but at this moment I see no other (backward compatible) solutions.

@jhillairet

**Comment by FranzForstmayr** [comment]
> An alternative option is to subclass Network into Network1port, Network2port, ..., NetworkManyPort and set the appropriate properties for each subclass.

It's not possible to choose a parent class dynamically. Apart from that it's not very beautiful.

> And a mitigation could be to modify the __dir__ of Network to filter out values of sm_n that are not valid for the number of ports.

This should work probably, but isn't a clean solution either I think. I'll start the investigate later today hopefully.

**Comment by FranzForstmayr** [comment]
> Since dir works on the instance we can modify that. The incorrect smn parameters are still part of the object, but they won't show any more in auto-completion. It is only a (partial) mitigation, but at this moment I see no other (backward compatible) solutions.

Does s11 and so on show up in autocompletion now?

**Comment by FranzForstmayr** [comment]
image

vscode fails to show completion

**Comment by eendebakpt** [comment]
In Jupyterlab the values do not show in autocompletion:

image

**Comment by jhillairet** [comment]
@FranzForstmayr @eendebakpt what about this PR?

**Comment by eendebakpt** [comment]
Closing in favor of #810

**Pull request closed without merging**